l1-contracts/src/core/libraries/staking/AddressSnapshotLib.sol

```diff
@@ -0,0 +1,226 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2025 Aztec Labs.
+pragma solidity >=0.8.27;
+
+import {Errors} from "@aztec/core/libraries/Errors.sol";
+import {Timestamp, Epoch, TimeLib} from "@aztec/core/libraries/TimeLib.sol";
+
+/**
+ * @notice Structure to store a set of addresses with their historical snapshots
+ * @param size The current number of addresses in the set
+ * @param checkpoints Mapping of index to array of address snapshots
+ * @param validatorToIndex Mapping of validator address to its current index in the set
+ */
+struct SnapshottedAddressSet {
+  uint256 size;
+  mapping(uint256 index => AddressSnapshot[]) checkpoints;
+  // Store up to date position for each validator
+  mapping(address validator => uint256 index) validatorToIndex;
```

Comment:
in future, remove this reverse lookup, and removing a validator requires also providing their index at the time - then check it there

Reply:
make issue